File: server/src/main/java/moe/shizuku/api/BinderContainer.java
package moe.shizuku.api;

import android.os.IBinder;
import android.os.Parcel;
import android.os.Parcelable;

/**
 * Shizuku 兼容的 BinderContainer
 * 包名必须是 moe.shizuku.api 才能被 Shizuku 应用正确反序列化
 */
public class BinderContainer implements Parcelable {

    public IBinder binder;

    public BinderContainer(IBinder binder) {
        this.binder = binder;
    }

    protected BinderContainer(Parcel in) {
        binder = in.readStrongBinder();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeStrongBinder(binder);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<BinderContainer> CREATOR = new Creator<BinderContainer>() {
        @Override
        public BinderContainer createFromParcel(Parcel in) {
            return new BinderContainer(in);
        }

        @Override
        public BinderContainer[] newArray(int size) {
            return new BinderContainer[size];
        }
    };
}

File: shizuku/api/src/main/kotlin/moe/shizuku/api/BinderContainer.kt
package moe.shizuku.api

import android.os.IBinder
import android.os.Parcel
import android.os.Parcelable

/**
 * Shizuku API 兼容类
 * 用于包装 IBinder 以便通过 ContentProvider 传递
 */
class BinderContainer(val binder: IBinder?) : Parcelable {

    constructor(parcel: Parcel) : this(parcel.readStrongBinder())

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeStrongBinder(binder)
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Parcelable.Creator<BinderContainer> {
        override fun createFromParcel(parcel: Parcel): BinderContainer {
            return BinderContainer(parcel)
        }

        override fun newArray(size: Int): Array<BinderContainer?> {
            return arrayOfNulls(size)
        }
    }
}
